src/Transform/ONNX/ShapeInferencePass.cpp

```diff
@@ -93,34 +97,45 @@ class ShapeInferencePass : public mlir::PassWrapper<ShapeInferencePass,
       signalPassFailure();
   }
 
-  static LogicalResult runShapeInferenceOnRegion(mlir::Region &r) {
+  static LogicalResult runShapeInferenceOnRegion(Region &r) {
+    std::function<void(Region &)> doShapeInference =
+        &ShapeInferencePass::runShapeInferenceOnRegion;
+
     // Iterate on the operations that need shape inference i.e the operations
     // that return a dynamic shape or followed by a return op.
     for (Operation &op : r.getOps()) {
-      std::function<void(mlir::Region &)> doShapeInference =
-          &ShapeInferencePass::runShapeInferenceOnRegion;
       // The shape of graph output has been imported from onnx protobuf model,
       // so the ops followed by a return op may not have dynamic shape output.
-      // However, shape inference is still need on these ops
-      // to infer optional attributes.
-      if (containSubgraph(&op) || isUsedByReturnOp(&op) ||
-          returnsDynamicOrUnknownShape(&op)) {
-        if (auto shape_op = llvm::dyn_cast<ShapeInference>(op)) {
-          if (failed(shape_op.inferShapes(doShapeInference))) {
-            op.emitError("shape inference failed");
-            return failure();
-          }
-        } else if (!llvm::dyn_cast<CallOpInterface>(op)) {
-          op.emitError("unable to infer shape of operation without shape "
-                       "inference interface");
-          return failure();
-        }
-      }
+      // However, shape inference is still needed on these ops to infer optional
+      // attributes.
+      if (!containSubgraph(op) && !isUsedByReturnOp(op) &&
+          !returnsDynamicOrUnknownShape(op))
+        continue;
+
+      if (auto shape_op = llvm::dyn_cast<ShapeInference>(op)) {
+        LLVM_DEBUG(
+            llvm::dbgs() << DEBUG_TYPE << ": op " << op.getName() << "\n");
+        // Verify the operation before attempting to infer the shape of the
+        // produced output(s).
+        Optional<RegisteredOperationName> registeredInfo =
+            op.getName().getRegisteredInfo();
+        if (registeredInfo && failed(registeredInfo->verifyInvariants(&op)))
```

Comment:
Does this `verifyInvariants` finally call our custom `verify` function? Or do we need to override it for each op?

Reply:
It does call the custom verify member function:
```
  static LogicalResult verifyInvariants(Operation *op) {
    static_assert(hasNoDataMembers(),
                  "Op class shouldn't define new data members");
    return failure(
        failed(op_definition_impl::verifyTraits<VerifiableTraitsTupleT>(op)) ||
        failed(cast<ConcreteType>(op).verify()));   <<<<<< HERE
  }
  ```

Reply:
Thanks for pointing it out!